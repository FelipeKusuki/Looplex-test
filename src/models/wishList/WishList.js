import { destroy, getParent, types } from "mobx-state-tree"

export const WishListItem = types.model({
    name: types.string,
    price: types.number,
    image: ""
})
.actions(obj => ({
    changeName(newName) {
        obj.name = newName
    },
    changePrice(newPrice) {
        obj.price = newPrice
    },
    changeImage(newImage) {
        obj.image = newImage
    },
    remove() {
        getParent(obj, 2).remove(obj)
    }
}))

export const WishList = types.model({
    items: types.optional(types.array(WishListItem), [])
})
.actions(obj => ({
    add(item) {
        obj.items.push(item)
    },
    remove(item) {
        destroy(item)
    }
}))
.views(obj => ({
    get totalPrice() {
        return obj.items.reduce((sum, entry) => sum + entry.price, 0)
    }
}))