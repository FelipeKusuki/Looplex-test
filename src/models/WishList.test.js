import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree"
import { reaction } from "mobx"
import { WishList, WishListItem } from "./WishList.js";

let item = null;
let list = null;

beforeEach(() => {
    item = WishListItem.create({
        name: "PlayStation 5",
        price: 5000
    })

    list = WishList.create({
        items: [
            {
                name: "Garrafa de Ã¡gua",
                price: 5.50
            }
        ]
    })
});

it("Cria instancia do model", () => {
    expect(item.price).toBe(5000);
    expect(item.image).toBe("");
})

it("Cria uma lista de desejos", () => {
    expect(list.items.length).toBe(1);
    expect(list.items[0].price).toBe(5.50);
})

it("Adicionando novos itens na lista com Patches", () => {
    const list = WishList.create()
    const patches = []
    onPatch(list, patch => {
        patches.push(patch)
    })
    list.add(item)
    list.items[0].changeName("PlayStation 5")
    expect(patches).toMatchSnapshot()
})

it("can add new items", () => {
    const list = WishList.create()
    const states = []
    onSnapshot(list, snapshot => {
        states.push(snapshot)
    })

    list.add(item)

    expect(list.items.length).toBe(1)
    expect(list.items[0].name).toBe("PlayStation 5")

    list.items[0].changeName("Xbox")

    expect(list.items[0].name).toBe("Xbox")
    expect(getSnapshot(list)).toMatchSnapshot()
    expect(states).toMatchSnapshot()
})
