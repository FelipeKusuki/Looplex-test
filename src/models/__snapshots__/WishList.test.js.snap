// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adicionando novos itens na lista com Patches 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "PlayStation 5",
      "price": 5000,
    },
  },
]
`;

exports[`can add new items 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "Xbox",
      "price": 5000,
    },
  ],
}
`;

exports[`can add new items 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "PlayStation 5",
        "price": 5000,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Xbox",
        "price": 5000,
      },
    ],
  },
]
`;
